### ISSUE TITLE
Performance of Tweedie log-likelihood

### ISSUE THREAD

**Issue by lorentzenchr** [Quantco/glum#721]
I know, this is not the primary focus of glum. But I thought it might be interesting:

```python
import numpy as np
from scipy.special import wright_bessel
from glum._distribution import TweedieDistribution

def tweedie_1(y, mu, p, phi):
    # We omit handling y==0 for brevity.
    theta = mu ** (1 - p) / (1 - p)
    kappa = mu ** (2 - p) / (2 - p)
    alpha = (2 - p) / (1 - p)
    ll_obs = (y * theta - kappa) / phi

    x = ((p - 1) * phi / y) ** alpha
    x /=  (2 - p) * phi
    wb = wright_bessel(-alpha, 0, x)
    ll_obs += np.log(wb / y)
    
    return np.sum(ll_obs)

def tweedie_2(y, mu, p, phi):
    td = TweedieDistribution(power=p)
    return td.log_likelihood(
        y=y,
        mu=mu,
        dispersion=phi,
    )

rng = np.random.default_rng(42)
n = 10_000
y = rng.gamma(2, scale=1.5, size=n)
mu = rng.gamma(2, scale=1.5, size=n)
p = 1.5
phi = 1.2

tweedie_1(y, mu, p, phi), tweedie_2(y, mu, p, phi)
# (-26732.28570241986, -26732.285702419867)

%time tweedie_1(y, mu, p, phi)
# CPU times: user 49.4 ms, sys: 1.46 ms, total: 50.9 ms
# Wall time: 50.2 ms

%time tweedie_2(y, mu, p, phi)
# CPU times: user 2.81 s, sys: 10.9 s, total: 13.7 s
# Wall time: 4.04 s
```

So it's almost a factor of 100. The problem with `wright_bessel` is that it is less robust especially for extreme parameters (e.g. internally exp(400) resulting in inf or nan).

**Comment by lbittarello** [comment]
Interesting. [We have had this discussion before](https://github.com/Quantco/glum/issues/481#issuecomment-970456385), but SciPy didn't have such an edge then. We could make the switch if we find that it's worth sacrificing robustness for performance. (@MarcAntoineSchmidtQC found that the Wight-Bessel function was unstable for Tweedie parameters close to 1 and 2 as well – see thread.)

**Comment by MarcAntoineSchmidtQC** [comment]
@lorentzenchr, the wall time you showed above is 99.9% used to compile the cython function. If you run it using the timeit cell magic instead of time, you will see that glum's tweedie distribution is 10x faster than wright_bessel. %time only runs the function once, while timeit runs it a number of times and takes the average.

Using your code above:

```python
%timeit tweedie_1(y, mu, p, phi)
# 25.1 ms ± 32.4 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

%timeit tweedie_2(y, mu, p, phi)
# 1.72 ms ± 7.94 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

**Comment by MarcAntoineSchmidtQC** [comment]
Sorry @lorentzenchr, I was using an environment with an old version of glum. At least we know this is a regression and it will be easier to debug! Thanks

**Comment by MarcAntoineSchmidtQC** [comment]
After a bit of investigating, the culprit seems to be cython 3. All releases before 2.6.0 were built using cython < 3, so they are not affected by this performance regression. Building the main branch using cython 0.29.36 leads to the same performance as before (~2ms)

I am not sure what to do. We can temporarily pin cython to < 3. Or we can further investigate and potentially raise an issue with cython. Because we are using cython in many different places, I would be surprised if the only performance regression was with the tweedie log likelihood. Any thoughts @jtilly, @xhochy?

**Comment by lorentzenchr** [comment]
Also scikit-learn and scipy noticed a performance regression with Cython 3. Fortunately, this seems to be fixed in Cython 3.0.3.

**Comment by MarcAntoineSchmidtQC** [comment]
Interesting. Thanks for the info! Seems that our performance regression is different. I tried with 3.0.3 and 3.0.5 and both lead to the same extremely slow code.
